### PR TITLE
ksh: deprecate

### PR DESCRIPTION
**Pull request by johnsonjh**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Deprecate `ksh` (which is actually [ksh2020](https://github.com/ksh2020/ksh)).

This project is no longer maintained and abandoned - a fork of the the unstable [ksh93v-](https://github.com/att/ast/tree/ksh93v). AT&T [no longer maintains](https://github.com/att/ast#readme) *any* version of [ksh](https://en.wikipedia.org/wiki/KornShell).  The *only* maintained ksh93 shell now is [ksh93u+m](https://github.com/ksh93/ksh/releases), and will be provided by a new `ksh93` formula.

The `ksh` formula switched from installing `ksh93u` to installing `ksh2020` in 2019.  `ksh2020` was an unsuccessful attempt to fork `ksh93` and breathe new life into the KornShell by extensively refactoring the unstable AST beta version. As that project has ceased all activity (approximately three years ago) this `ksh` formula will never receive any bug-fixes, security fixes, or other any other updates and must (eventually) be removed from Homebrew. 

Users should install the up-to-date and maintained ksh93u+m shell via the `ksh93` formula (and not `ksh`.)

### CODE REVIEW

**Comment by carlocab** [comment]
Please actually follow the commit style guide, especially if you tick it as completed.

Here's a quick summary for you.

For new formulae:

> At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.

For existing formulae:

> The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

**Comment by johnsonjh** [comment]
Sorry about breaking the norms, I actually did think I was being helpful by giving the context up front that `ksh` itself is not actually being deprecated but only the `ksh` formula, and that it is actually `ksh2020` and not `ksh93` (since that distinction has become important now), yadda yadda... The whole situation is super confusing.  I'll be more pedantic going forward.

I also sincerely apologize if I come off as some obsessive lunatic about `ksh`. For interactive use, `zsh` might be OK, but a working `ksh` is something I can't live without, at least without literally ending up as a lunatic ... "*using `ksh`*" is synonymous with "*using computers*" for me, and has been for more than three decades.

So, if `ksh` is broken, then I'm broken too. 😿

**Comment by carlocab** [comment]
You are free to add whatever you want after the header
```
ksh: deprecate
```
The header you initially added will just get swallowed up by @BrewTestBot and turned into something uninformative like
```
ksh: rebuild
```
or something.

You can avoid that by actually following the format, so your commit message looks something like
```
ksh: deprecate


Some historical notes... etc... etc...
```

**Comment by johnsonjh** [comment]
@carlocab 

If possible, I'd like to advocate to **outright disable this formula if the new ksh93 formula is accepted**.

I've lost untold hours of work and even crashed the entire laptop system, just from trying to use ksh normally.  Yes - it's *really that bad*.

Here are the two most egregious examples out of the dozens and dozens of problems in ksh as shipped by Homebrew, all of which I could document with ease - I easily fill hundreds of pages of text describing all the problems in ksh2020! Below, I will show comparisons against the ksh version that's shipped by everyone else. I used RHEL for the lower half of these screenshots.

#### Showstopper 1 - History totally broken

This absolutely kills any possible interactive usage whatsoever of this shell.

##### Steps to reproduce

1) Setup standard recall, map the up arrow to last command recall (`^P`).
2) Type a continued/compound command, but make a mistake - I forget to add `do`, and I get an error.
3) Press `^P` (or `Up`) for recall, so you can fix the problem.

##### Results

* In ksh93, using `^P` (or `Up`) puts me back at the last valid part of the command. I can continue from the point where I made the error, as expected.
* In ksh2020, hilarity ensues; you can't recall any part of the command you mistyped at all, and you need to retype everything over again.

##### Screenshot

Screenshot from 2022-08-02 23-01-38

#### Showstopper 2 - ksh2020 goes haywire, eats all every available file descriptor.

* ***TL;DR - Literally everything breaks every 200-300 commands you type.*** No joke.

I have no idea how this ever got in a "released" shell or what the root cause is, but you can reproduce it trivially.

It seems related to the history again, triggered if you have *any* past history loaded from disk. Granted, I normally use a rather small history buffer, and that makes the problem happen almost immediately. 

Even in the out-of-box default configuration, typing only a few hundred shell commands will kill your session if not your system.  A ksh user will hit this bug dozens of times every single day.

##### Steps to reproduce

* Set `PS1='[!]\$ '` so you can see the history count, and,
  1. Just use the shell normally until everything crashes and burns, or,
  2. With `HISTSIZE=30` just to make it faster to show the issue - it happens with any HISTSIZE, eventually! - type a few hundred regular commands, and exit ksh and then start it again.

##### Results

* ksh2020's history system completely fails to function, no entries are added, etc.
* ksh93 works fine as usual.

##### Screenshot

Screenshot from 2022-08-02 23-20-40

* **BONUS!** As if this wasn't bad enough, if you dare to continue to use the shell in this state, it somehow randomly starts leaking open file descriptors at an insane rate - a resource the MacOS limits to 256 or so, by default.

* You'll quickly exhaust them all and now eventually everything starts spewing errors like "too many open files".  You'll need to kill your session if it's via SSH and hope the Mac cleans up, and possibly even reboot your computer if your using it directly.  Fail!

**Review by SMillerDev — APPROVED** [review]
